File: sdks/python/src/splyntra/instrumentors/browser_use.py
# SPDX-License-Identifier: Apache-2.0
"""Browser Use auto-instrumentor — instruments browser-use agents, steps, and browser tool actions.

Emits:
- A root ``agent`` span for ``Agent.run()``
- A ``step`` span for each ``Agent.step()``
- A ``tool_call`` span for each browser controller action (navigate, click, type, extract, screenshot)
"""

from __future__ import annotations

import functools
import time
from typing import Any, Collection

from opentelemetry import trace
from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
from opentelemetry.trace import StatusCode

_FRAMEWORK = "browser-use"


class BrowserUseInstrumentor(BaseInstrumentor):
    """Instruments browser-use Agent execution and controller actions.

    Usage:
        from splyntra.instrumentors import BrowserUseInstrumentor
        BrowserUseInstrumentor().instrument()
    """

    def instrumentation_dependencies(self) -> Collection[str]:
        from importlib.util import find_spec

        if find_spec("browser_use") is not None:
            return ("browser-use >= 0.1.0",)
        return ()

    def _instrument(self, **kwargs):
        tracer = trace.get_tracer("splyntra.browser_use")
        self._patched: list[tuple[Any, str, Any]] = []

        # 1. Patch Agent.run (async)
        self._patch_agent_run(tracer)

        # 2. Patch Agent.step (async)
        self._patch_agent_step(tracer)

        # 3. Patch Controller actions
        self._patch_controller_act(tracer)

    def _patch_agent_run(self, tracer: trace.Tracer):
        try:
            from browser_use.agent.service import Agent
        except Exception:  # noqa: BLE001
            try:
                from browser_use import Agent
            except Exception:  # noqa: BLE001
                return

        original = getattr(Agent, "run", None)
        if original is None or getattr(original, "_splyntra_wrapped", False):
            return

        @functools.wraps(original)
        async def wrapped_run(self_agent, *args, **kwargs):
            task_desc = getattr(self_agent, "task", None) or "browser_task"
            span_name = f"browser_use.agent:{task_desc[:40]}" if isinstance(task_desc, str) else "browser_use.agent"

            with tracer.start_as_current_span(
                span_name,
                kind=trace.SpanKind.INTERNAL,
                attributes={
                    "splyntra.span.type": "agent",
                    "splyntra.framework": _FRAMEWORK,
                    "splyntra.agent.name": "browser_agent",
                    "splyntra.workflow": "web_browsing",
                    "browser.task": str(task_desc)[:500],
                },
            ) as span:
                start = time.perf_counter()
                try:
                    history = await original(self_agent, *args, **kwargs)
                    span.set_attribute("splyntra.tool.duration_ms", (time.perf_counter() - start) * 1000)
                    span.set_status(StatusCode.OK)
                    return history
                except Exception as e:  # noqa: BLE001
                    span.set_status(StatusCode.ERROR, str(e))
                    span.record_exception(e)
                    raise

        setattr(wrapped_run, "_splyntra_wrapped", True)
        setattr(Agent, "run", wrapped_run)
        self._patched.append((Agent, "run", original))

    def _patch_agent_step(self, tracer: trace.Tracer):
        try:
            from browser_use.agent.service import Agent
        except Exception:  # noqa: BLE001
            try:
                from browser_use import Agent
            except Exception:  # noqa: BLE001
                return

        original = getattr(Agent, "step", None)
        if original is None or getattr(original, "_splyntra_wrapped", False):
            return

        @functools.wraps(original)
        async def wrapped_step(self_agent, *args, **kwargs):
            step_num = getattr(self_agent, "n_steps", 0)
            span_name = f"browser_use.step_{step_num}"

            with tracer.start_as_current_span(
                span_name,
                kind=trace.SpanKind.INTERNAL,
                attributes={
                    "splyntra.span.type": "step",
                    "splyntra.framework": _FRAMEWORK,
                    "splyntra.step.name": span_name,
                    "browser.step_number": step_num,
                },
            ) as span:
                start = time.perf_counter()
                try:
                    result = await original(self_agent, *args, **kwargs)
                    span.set_attribute("splyntra.tool.duration_ms", (time.perf_counter() - start) * 1000)
                    span.set_status(StatusCode.OK)
                    return result
                except Exception as e:  # noqa: BLE001
                    span.set_status(StatusCode.ERROR, str(e))
                    span.record_exception(e)
                    raise

        setattr(wrapped_step, "_splyntra_wrapped", True)
        setattr(Agent, "step", wrapped_step)
        self._patched.append((Agent, "step", original))

    def _resolve_tool_classes(self) -> list:
        """Resolve the action-executor class across browser-use versions.

        browser-use renamed ``Controller`` → ``Tools`` in 0.3.x (``Controller`` is
        kept as a backward-compatible alias in most releases). We patch whichever
        class(es) are importable so ``tool_call`` spans keep flowing on both the
        legacy and current API — de-duped by identity so an alias isn't wrapped
        twice.
        """
        import importlib

        candidates = (
            ("browser_use.controller.service", "Controller"),
            ("browser_use.tools.service", "Tools"),
            ("browser_use", "Controller"),
            ("browser_use", "Tools"),
        )
        seen: set[int] = set()
        classes: list = []
        for module_name, attr in candidates:
            try:
                mod = importlib.import_module(module_name)
            except Exception:  # noqa: BLE001
                continue
            cls = getattr(mod, attr, None)
            if cls is not None and id(cls) not in seen:
                seen.add(id(cls))
                classes.append(cls)
        return classes

    def _patch_controller_act(self, tracer: trace.Tracer):
        tool_classes = self._resolve_tool_classes()
        if not tool_classes:
            return

        def make_wrapper(orig):
            @functools.wraps(orig)
            async def wrapped_act(self_ctrl, action, *args, **kwargs):
                action_name = getattr(action, "__class__", type(action)).__name__
                if hasattr(action, "model_dump"):
                    action_data = str(action.model_dump())[:400]
                else:
                    action_data = str(action)[:400]

                span_name = f"browser.{action_name.lower()}"
                with tracer.start_as_current_span(
                    span_name,
                    kind=trace.SpanKind.INTERNAL,
                    attributes={
                        "splyntra.span.type": "tool_call",
                        "splyntra.tool.name": span_name,
                        "splyntra.framework": _FRAMEWORK,
                        "splyntra.input": action_data,
                    },
                ) as span:
                    start = time.perf_counter()
                    try:
                        res = await orig(self_ctrl, action, *args, **kwargs)
                        span.set_attribute("splyntra.tool.duration_ms", (time.perf_counter() - start) * 1000)
                        if res is not None:
                            span.set_attribute("splyntra.output", str(res)[:1000])
                        span.set_status(StatusCode.OK)
                        return res
                    except Exception as e:  # noqa: BLE001
                        span.set_status(StatusCode.ERROR, str(e))
                        span.record_exception(e)
                        raise

            return wrapped_act

        for cls in tool_classes:
            for method in ("act", "multi_act"):
                original = cls.__dict__.get(method)
                if original is None or getattr(original, "_splyntra_wrapped", False):
                    continue

                wrapped = make_wrapper(original)
                setattr(wrapped, "_splyntra_wrapped", True)
                setattr(cls, method, wrapped)
                self._patched.append((cls, method, original))

    def _uninstrument(self, **kwargs):
        for target, attr, original in getattr(self, "_patched", []):
            setattr(target, attr, original)
        self._patched = []

File: sdks/typescript/src/instrumentors/browser-use.ts
// SPDX-License-Identifier: Apache-2.0
import { trace, SpanKind, SpanStatusCode, Span } from "@opentelemetry/api";
import { patchDual, pick } from "./patch";

/**
 * Auto-instrument Browser Use for JavaScript / TypeScript.
 *
 * Wraps:
 * - Agent.prototype.run (agent root span)
 * - Agent.prototype.step (step span)
 * - Controller.prototype.act / multiAct (tool_call span)
 *
 * Best-effort: a safe no-op if the package is absent or its API differs.
 * Patches both CJS and ESM module exports.
 *
 * Returns true if the patch wrapped at least one method.
 */
export function instrumentBrowserUse(): boolean {
  return patchDual(["browser-use", "@browser-use/core", "@browser-use/agent"], patchBrowserUse);
}

function patchBrowserUse(mod: unknown): boolean {
  const Agent = pick(mod, "Agent") as { prototype?: any } | undefined;
  const Controller = pick(mod, "Controller") as { prototype?: any } | undefined;

  let wrapped = false;

  // 1. Wrap Agent.prototype.run
  wrapped =
    wrapMethod(
      Agent?.prototype,
      ["run", "runAsync", "execute"],
      "agent",
      (self, args) => {
        const task = self?.task || args?.[0] || "browser_task";
        const taskSnippet = typeof task === "string" ? task.slice(0, 40) : "browser_task";
        return `browser_use.agent:${taskSnippet}`;
      },
      (self, args) => ({
        "splyntra.framework": "browser-use",
        "splyntra.agent.name": "browser_agent",
        "splyntra.workflow": "web_browsing",
        "browser.task": String(self?.task || args?.[0] || "").slice(0, 500),
      })
    ) || wrapped;

  // 2. Wrap Agent.prototype.step
  wrapped =
    wrapMethod(
      Agent?.prototype,
      ["step", "stepAsync"],
      "step",
      (self) => {
        const stepNum = self?.n_steps ?? self?.stepCount ?? 0;
        return `browser_use.step_${stepNum}`;
      },
      (self) => ({
        "splyntra.framework": "browser-use",
        "browser.step_number": self?.n_steps ?? self?.stepCount ?? 0,
      })
    ) || wrapped;

  // 3. Wrap Controller.prototype.act
  wrapped =
    wrapMethod(
      Controller?.prototype,
      ["act", "multiAct", "runAction"],
      "tool_call",
      (_self, args) => {
        const action = args?.[0];
        const actionName = action?.constructor?.name || action?.name || action?.type || "action";
        return `browser.${String(actionName).toLowerCase()}`;
      },
      (_self, args) => {
        const action = args?.[0];
        const inputData = action ? JSON.stringify(action).slice(0, 400) : "";
        return {
          "splyntra.framework": "browser-use",
          "splyntra.input": inputData,
        };
      }
    ) || wrapped;

  return wrapped;
}

function wrapMethod(
  proto: any,
  methodNames: string[],
  spanType: string,
  nameOf: (self: any, args: any[]) => string,
  attributesOf?: (self: any, args: any[]) => Record<string, any>
): boolean {
  if (!proto) return false;
  const method = methodNames.find((m) => typeof proto[m] === "function" && !proto[m].__splyntraWrapped);
  if (!method) return false;
  const original = proto[method];

  function patched(this: any, ...args: any[]) {
    const name = nameOf(this, args);
    const extraAttrs = attributesOf ? attributesOf(this, args) : {};
    const tracer = trace.getTracer("splyntra.browser-use");

    return tracer.startActiveSpan(
      name,
      {
        kind: SpanKind.INTERNAL,
        attributes: {
          "splyntra.span.type": spanType,
          [spanType === "tool_call" ? "splyntra.tool.name" : spanType === "step" ? "splyntra.step.name" : "splyntra.agent.name"]: name,
          "splyntra.framework": "browser-use",
          ...extraAttrs,
        },
      },
      (span: Span) => {
        let result: any;
        const start = performance.now();
        try {
          result = original.apply(this, args);
        } catch (err: any) {
          span.setAttribute("splyntra.tool.duration_ms", performance.now() - start);
          span.setStatus({ code: SpanStatusCode.ERROR, message: err?.message });
          span.recordException(err);
          span.end();
          throw err;
        }

        return Promise.resolve(result)
          .then((res: any) => {
            span.setAttribute("splyntra.tool.duration_ms", performance.now() - start);
            if (res !== undefined && res !== null) {
              span.setAttribute("splyntra.output", typeof res === "string" ? res.slice(0, 1000) : JSON.stringify(res).slice(0, 1000));
            }
            span.setStatus({ code: SpanStatusCode.OK });
            span.end();
            return res;
          })
          .catch((err: any) => {
            span.setAttribute("splyntra.tool.duration_ms", performance.now() - start);
            span.setStatus({ code: SpanStatusCode.ERROR, message: err?.message });
            span.recordException(err);
            span.end();
            throw err;
          });
      }
    );
  }

  (patched as any).__splyntraWrapped = true;
  proto[method] = patched;
  return true;
}
